### PR TITLE
i18n integration.

### LINKED ISSUE #84: Add i18n
Using module https://github.com/besport/ocsigen-i18n
For example, translate all error messages ("wrong password", etc) in french.
Also detect the language from browser settings or mobile device (ask Besport about that).

### PR DESCRIPTION
**Pull request by dannywillems**
Fix #84.

## Dependency.
- [x] ocsigen-i18n: https://github.com/besport/ocsigen-i18n/pull/17 must be merged first.
- [ ] add ocsigen-i18n in OPAM, README, Jenkins, Travis.

## Manual.
- [ ] Write a manual.

## Template:
- [x] Add in makefiles the generation of i18n file.
- [x] Add a tsv sample file. Added in assets directory in the template. **Text must be added. See `templates.distillery/assets!PROJECT_NAME_i18n.tsv`.**
- [x] Use the PPX extension for the translation.
- [x] Add a language template file to get a sum type with languages (see https://github.com/besport/ocsigen-i18n) --> `PROJECT_NAME_language.eliom[i]`.
- [x] Update SQL file by adding a field `language` in `ocsigen_start.users` table.
- [x] Documentation, documentation and documentation!!!
- [ ] Test, test and test!!!
## Library:
- [x] Add a field language in the `ocsigen_start.users` tables.
- [x] Update functions which access to users table.
- [x] Update type `Os_types.users`.
- [x] Detect the user language when starting a process. If not connected, based on Accept-Language value in the HTTP header. If no value is given in the header, the default value is `"en-US"` and it can be changed with `Os_language.set_default_language`. If the user is connected, it retrieves the language from the database.
- [x] `Os_userbox` and `Os_handlers` depend on some translations, which are not available in the library (because only available in the template. Even if it's ugly, `Os_i18n` contains references to the text and they are overwritten by the template for the appropriate translations (see `PROJECT_NAME_language`).
## Arbitrary choice
- The language can be NULL in the table. It implies:
  - When creating/updating an user, the language is optional. See `Os_db.User.update` and `Os_db.User.create`. 
- `set_personal_data` doesn't require the language. The language is not considered as a personal data.
## Issue.
- [x] `eliomdep` is executed before `make i18n-update`

### CODE REVIEW

**Comment by dannywillems** [comment]
I can't run the server. I have the fatal error:
```
ocsigenserver: ocsigen:main: Fatal - Error in configuration file: Error while parsing configuration file: Eliom: while loading local/lib/test_i18n/test_i18n.cma: (Failure
ocsigenserver: ocsigen:main:   "This function cannot be called here because it needs information about the request or the site.")
```
I can't find the line which is causing the issue :(.
@vasilisp: any idea?
You need to pin https://https://github.com/sagotch/ocsigen-i18n/tree/type-t (would be merged soon).

**Comment by vasilisp** [comment]
@dannywillems there are quite a few things that you can only do during a request, so hard to guess. Do you know the commit that caused this?

**Comment by dannywillems** [comment]
@vasilisp I don't think so. As it will be rebased, I commit by files. Lots of changes is about text.

**Comment by dannywillems** [comment]
The last thing is about testing and find this error. Else, I think it's finished.

I don't like the use of `Os_i18n` because it's not elegant that all.

File: src/os_i18n.eliom [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+
+let%shared passwords_do_not_match =
+  ref "Passwords do not match"
```
Comment (vasilisp) [inline src/os_i18n.eliom]: I haven't followed everything, but I don't understand how such server-side references work. If you change the contents, that will impact all users.

Comment (dannywillems) [inline src/os_i18n.eliom]: Oh... You're right... :confused:

**Comment by dannywillems** [comment]
The error comes from the `%%%PROJECT_NAME%%%_language` when the translation is done (variables are not defined...).

File: template.distillery/PROJECT_NAME_language.eliom [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            string_of_language %%%MODULE_NAME%%%_i18n.default_language));
+        %%%MODULE_NAME%%%_i18n.default_language)
+      in
+      set_language language;
+      ignore ([%client (set_language ~%language : unit)]);
+      Lwt.return ())
+
+let%shared _ =
+  Os_i18n.passwords_do_not_match :=
+    [%i18n S.passwords_do_not_match];
```
Comment (dannywillems) [inline template.distillery/PROJECT_NAME_language.eliom]: I think the error comes from here when the translation is done (variables are not defined...).

**Comment by dannywillems** [comment]
I wait #236 to rebase and continue because it will have new conflicts.

**Comment by dannywillems** [comment]
@balat Do I wait you finish the template to rebase and continue i18n integration?

**Comment by balat** [comment]
Hello @dannywillems 
The template won't change a lot now. Do you plan to finish this before next week ?
Otherwise it will be for next version.

**Comment by dannywillems** [comment]
@balat Yes I will do it. About translations in the library, I will use your idea with functions.

**Comment by dannywillems** [comment]
I have still no idea where the error comes from :disappointed:.

**Comment by dannywillems** [comment]
As I can't resolve the issue, I retry step by step. And I don't want to resolve all conflicts :D. See #361

**Pull request closed without merging**